From c9144de9f2d1512eb319c74163657ab4f2502786 Mon Sep 17 00:00:00 2001
From: Lu Jialin <lujialin4@huawei.com>
Date: Sat, 28 Jan 2023 18:32:09 +0800
Subject: [PATCH 052/905] [Huawei] memcg: export high_async_ratio to userland

Offering: HULK
hulk inclusion
category: feature
bugzilla: 188244

-------------------------------

User can set high_async_ratio from 10 to 100; start memcg high async when
memcg_usage is larger than memory.high * high_async_ratio / 100;

Signed-off-by: Lu Jialin <lujialin4@huawei.com>
---
 mm/memcontrol.c | 38 ++++++++++++++++++++++++++++++++++++++
 1 file changed, 38 insertions(+)

diff --git a/mm/memcontrol.c b/mm/memcontrol.c
index 6a8439dc36ff..22f27bb49f34 100644
--- a/mm/memcontrol.c
+++ b/mm/memcontrol.c
@@ -5319,6 +5319,38 @@ static ssize_t wb_blkio_write(struct kernfs_open_file *of, char *buf,
 }
 #endif
 
+#ifdef CONFIG_MEMCG_V1_THRESHOLD_QOS
+static int memcg_high_async_ratio_show(struct seq_file *m, void *v)
+{
+	seq_printf(m, "%d\n",
+		   READ_ONCE(mem_cgroup_from_seq(m)->high_async_ratio));
+	return 0;
+}
+
+static ssize_t memcg_high_async_ratio_write(struct kernfs_open_file *of,
+					    char *buf, size_t nbytes, loff_t off)
+{
+	struct mem_cgroup *memcg = mem_cgroup_from_css(of_css(of));
+	int ret, high_async_ratio;
+
+	buf = strstrip(buf);
+	if (!buf)
+		return -EINVAL;
+
+	ret = kstrtoint(buf, 0, &high_async_ratio);
+	if (ret)
+		return ret;
+
+	if (high_async_ratio > HIGH_ASYNC_RATIO_BASE ||
+	    high_async_ratio <= HIGH_ASYNC_RATIO_GAP)
+		return -EINVAL;
+
+	WRITE_ONCE(memcg->high_async_ratio, high_async_ratio);
+
+	return nbytes;
+}
+#endif
+
 static struct cftype mem_cgroup_legacy_files[] = {
 	{
 		.name = "usage_in_bytes",
@@ -5513,6 +5545,12 @@ static struct cftype mem_cgroup_legacy_files[] = {
 		.file_offset = offsetof(struct mem_cgroup, events_local_file),
 		.seq_show = memcg_events_local_show,
 	},
+	{
+		.name = "high_async_ratio",
+		.flags = CFTYPE_NOT_ON_ROOT,
+		.seq_show = memcg_high_async_ratio_show,
+		.write = memcg_high_async_ratio_write,
+	},
 #endif
 
 #ifdef CONFIG_MEMCG_RECLAIM
-- 
2.17.1

